File: android/measure/src/main/java/sh/measure/android/bugreport/ShakeBugReportCollector.kt
package sh.measure.android.bugreport

import sh.measure.android.Measure
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference

internal class ShakeBugReportCollector(
    private val autoLaunchEnabled: Boolean,
    private val shakeDetector: ShakeDetector,
) : ShakeDetector.Listener {
    private val listener = AtomicReference<MsrShakeListener?>(null)
    private val autoLaunch = AtomicBoolean(false)
    private var takeScreenshot = false

    init {
        if (autoLaunchEnabled) {
            enableAutoLaunch(takeScreenshot)
        }
    }

    fun enableAutoLaunch(takeScreenshot: Boolean) {
        autoLaunch.set(true)
        this.takeScreenshot = takeScreenshot
        shakeDetector.setShakeListener(this)
        shakeDetector.start()
    }

    fun disableAutoLaunch() {
        autoLaunch.set(false)
        shakeDetector.setShakeListener(null)
        shakeDetector.stop()
    }

    fun setShakeListener(listener: MsrShakeListener?) {
        if (autoLaunch.get()) {
            return
        }
        this.listener.set(listener)
        if (listener == null) {
            shakeDetector.setShakeListener(null)
            shakeDetector.stop()
        } else {
            shakeDetector.setShakeListener(this)
            shakeDetector.start()
        }
    }

    override fun onShake() {
        if (autoLaunch.get()) {
            Measure.launchBugReportActivity(takeScreenshot = takeScreenshot)
        } else {
            listener.get()?.onShake()
        }
    }

    fun isShakeToLaunchBugReportEnabled(): Boolean {
        return autoLaunch.get()
    }
}

File: ios/Sources/MeasureSDK/Swift/BugReport/ShakeBugReportCollector.swift
//
//  ShakeBugReportCollector.swift
//  Measure
//
//  Created by Adwin Ross on 09/05/25.
//

import Foundation
import UIKit

final class ShakeBugReportCollector: ShakeDetectorListener {
    private let shakeDetector: ShakeDetector
    private var shakeHandler: (() -> Void)?

    init(shakeDetector: ShakeDetector) {
        self.shakeDetector = shakeDetector
    }

    func setShakeHandler(_ handler: (() -> Void)?) {
        self.shakeHandler = handler
        if handler == nil {
            shakeDetector.setShakeListener(nil)
            shakeDetector.stop()
        } else {
            shakeDetector.setShakeListener(self)
            _ = shakeDetector.start()
        }
    }

    func onShake() {
        shakeHandler?()
    }
}
